Added OpenACC directives and Role checks, to enable running on CPU with other Physics on GPUs.

cponder opened the pull request:
Added OpenACC directives from 6.x-openacc version to manage Host <-> GPU memory movement, to work with some physics models running on the GPU.

This particular file, from my manual merge, came out the same as what was in mgduda's original 6.x -> develop merge.

I have successfully built & run the MPAS-A develop-openacc branch with this modifed file, with these two configurations:
1. CPU-only build, run with full physics simulation (including this modified file).
2. OpenACC build, but only the moist DyCore part of the simulation. This file is not tested in the run, but is show to compile OK with the changes.

raghuraj19 reviewed (approved):
This looks good to me.

raghuraj19 commented:
@cponder : Can you add @supreethms1809 @Pranay-Reddy-Kommera @cenamiller  and @mgduda as reviewers to this? That way we can begin merging this fast.

cponder commented:
@supreethms1809 
@Pranay-Reddy-Kommera
@cenamiller
@mgduda
I would like to add you all as reviewers but can't find a button to do this. Can you leave comments at least?

supreethms1809 commented:
@cponder

Pranay-Reddy-Kommera commented:
The changes look good to me.

On Wed, Jul 21, 2021 at 3:18 PM Carl Ponder ***@***.***>
wrote:

> @supreethms1809 <https://github.com/supreethms1809>
> @Pranay-Reddy-Kommera <https://github.com/Pranay-Reddy-Kommera>
> @cenamiller <https://github.com/cenamiller>
> @mgduda <https://github.com/mgduda>
> I would like to add you all as reviewers but can't find a button to do
> this. Can you leave comments at least?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MPAS-Dev/MPAS-Model/pull/893#issuecomment-884506374>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADY2MNWD6CIDHULKVY2CADDTY42Q3ANCNFSM5AQ64YIA>
> .
>


-- 

With Regards,
Pranay Reddy Kommera
***@***.***,
+1-307-399-3046

mgduda reviewed (changes requested):
@cponder Can you please change the commit message and the PR description to describe what the changes in this commit/PR actually accomplish? I'd also advocate for *not* referring to other developers by name (especially just first name) in commit messages, as it probably isn't helpful to anyone else who isn't on a first-name basis with the named developer. Saying something like "These changes come from Michael" doesn't really help anyone else understand the motivation behind the code changes.

cponder commented:
@mgduda -- I made some changes to the comment at the top.

@raghuraj19 -- I'm not sure how to explain what the changes really accomplish, given that it's really just a merge of the way the two files work.

cponder commented:
@mgduda I changed the title per Raghu's recommendation and added moved some details to the comment portion.

@raghuraj19 I changed the ROLE tests in a second commit, I assume this PR will include both commits.

mgduda commented on src/core_atmosphere/physics/mpas_atmphys_driver_cloudiness.F:
```diff
@@ -106,15 +109,21 @@ subroutine cloudiness_from_MPAS(configs,mesh,diag_physics,sfc_input,its,ite)
+ call mpas_pool_get_array_gpu(sfc_input,'xland',xland)
+ coupler => areaCellField%block%domain%mpas_cpl
+ if (coupler%role_includes(ROLE_INTEGRATE)) then
+        !$acc update host(areaCell, meshDensity, xland)
```
Why do we need to update `areaCell`, `meshDensity`, and `xland` on the host in every call to `cloudiness_from_MPAS`?

Pranay-Reddy-Kommera replied:
The update to the host is necessary as the next steps involve CPU code. As we are not tracking which individual variable is modified on GPU in each timestep, we try to update all the variables on host before the start of the CPU code. 
Finally, these update directives will be removed once the entire cloudiness scheme is ported onto GPU.

mgduda replied:
The `areaCell` and `meshDensity` fields don't change in time, and since these are first read from the model ICs by code executing on the CPU (if I understand the program flow correctly), I was just wondering why these updated were needed at all. Did someone find that these were required for correct results at some point in the past? If the plan is to later remove unnecessary transfers, then I won't object to adding these temporarily.

Part of the reason why I've been insisting on commit messages that actually describe what the code changes are doing and why is so that others can understand issues like the one discussed here.

Pranay-Reddy-Kommera replied:
We did not find that these were required for correct results. During MPAS 6.x Physics porting, we were not sure about which variable are updated on CPU/GPU, and keeping track of them during the porting would have taken more time. Moreover as these update directives will be removed once the entire scheme is ported, we updated all the variables for time-being.
Except in radiation related subroutines, all the update host in other schemes will be removed entirely. And for radiation, I believe we can later track and remove the unnecessary variables.

raghuraj19 replied:
@mgduda: Yes, these will be removed once the porting has been completed. It is just a precautionary step from our side to avoid any unanticipated bugs into the code. So, in the end, you will not see such updates in the stable versions of openacc code as seen in 6.x. These updates, though unnecessary, will be useful when we hit a bug. It allows to easily track down a variable that is deviating in results. So, would it be ok to keep this temporarily?

mgduda replied:
@raghuraj19 and @Pranay-Reddy-Kommera That makes sense -- thanks very much for the explanation!

mgduda commented:
@cponder Before we merge this, I would propose that we squash the two commits in this PR into a single commit with a message that describes what's going on in the code changes. I'm willing to handle the reworking of the branch, and even rewriting the commit message, but I think I'll need confirmation of a few details:
1. The cloud fraction scheme ("cloudiness") is going to be running on the CPU, which is why we have `update host` directives in `cloudiness_from_MPAS` and `update device` directives in `cloudiness_to_MPAS`?
2. The updates of the `areaCell`, `meshDensity`, and `xland` fields are anticipated to be removed eventually, but are added here as a precaution, while the update of the `cldfrac` field is considered required for correctness?

mgduda commented on src/core_atmosphere/physics/mpas_atmphys_driver_cloudiness.F:
```diff
@@ -121,7 +121,7 @@ subroutine cloudiness_from_MPAS(configs,mesh,diag_physics,sfc_input,its,ite)
  call mpas_pool_get_array_gpu(sfc_input,'xland',xland)
  coupler => areaCellField%block%domain%mpas_cpl
- if (coupler%role_is(ROLE_INTEGRATE)) then
+ if (coupler%role_includes(ROLE_INTEGRATE)) then
```
Is there really any sense in changing this from `role_is` to `role_includes`? If the role of a task is not *exclusively* integration, then it means the task is also handling radiation and cloud fraction calculations, in which case it doesn't seem like any update of fields should be needed. Did you find a specific need to change the conditional here to use `role_includes`?

raghuraj19 replied:
That was something I recommended. Since there is a chance that we will have the rrtmgf being introduced, we may need to run the radiation on the same roles as GPU. In case of a debug, if we ran the simulation without specifying the roles (ROLE03), then role_is would decline the condition but role_includes would pass the condition to keep cloudiness on the CPU. Am I right? That is why I asked for the switch.

mgduda replied:
I think that makes sense. Thanks for the explanation!

cponder commented:
@mgduda I'm fine with combining the commits, I just don't know how to deal with these things in this framework.

@raghuraj19  @Pranay-Reddy-Kommera   Can you answer MIchael's #1 & #2 from two messages ago?

Pranay-Reddy-Kommera commented:
Response to @mgduda questions.
1. The cloud fraction scheme ("cloudiness") is going to be running on the CPU, which is why we have update host directives in cloudiness_from_MPAS and update device directives in cloudiness_to_MPAS? Response: Yes

2. The updates of the areaCell, meshDensity, and xland fields are anticipated to be removed eventually, but are added here as a precaution, while the update of the cldfrac field is considered required for correctness? Response: Yes, eventually update of all the variables will be removed once the cloudiness is ported onto GPU.

supreethms1809 reviewed (approved):
Changes looks good to me. works on both casper and cheyenne with intel/nvhpc (on CPUs)

mgduda commented:
@cponder I've just pushed changes to rework the branch history and commit message.
